app/MMB/src/scenes/timeline/Timeline.js

```diff
@@ -53,35 +54,47 @@ function getValidTimelineEvent(data) {
   return null;
 }
 
-function renderDaySeparator(event, prevEvent) {
-  if (event && event.timestamp && prevEvent && prevEvent.timestamp) {
-    const eventTimestamp = new Date(event.timestamp);
-    const prevEventTimestamp = new Date(prevEvent.timestamp);
-    if (!DateUtils.isSameDay(eventTimestamp, prevEventTimestamp)) {
-      return <DaySeparator date={eventTimestamp} />;
-    }
-  }
-  return null;
-}
-
 function renderChildren(events) {
   if (events) {
-    return events.map((event, index) => {
-      if (event && event.timestamp && event.__typename) {
-        let daySeparator;
-        if (index === 0) {
-          daySeparator = <DaySeparator date={new Date(event.timestamp)} />;
-        } else {
-          daySeparator = renderDaySeparator(event, events[index - 1]);
-        }
-        return (
-          <React.Fragment key={'TimelineEvent-' + event.__typename + index}>
-            {daySeparator}
-            {getValidTimelineEvent(event)}
-          </React.Fragment>
+    const renderedEvents = events
+      .map(event => getValidTimelineEvent(event))
+      // Necessary to filter out non-implemented components which could appear in GQL response
+      // + to get correct day separators
+      // + to correctly determine the first and last events being displayed
+      .filter(renderedEvent => renderedEvent);
+    const timestamps = renderedEvents.map(renderedEvent =>
+      idx(renderedEvent, _ => _.props.data.timestamp),
+    );
+    return renderedEvents.map((renderedEvent, index) => {
+      const currentTimestamp = idx(renderedEvent, _ => _.props.data.timestamp);
+      let daySeparator = null;
+      if (currentTimestamp) {
+        const sameDays = timestamps.map(
+          timestamp =>
+            timestamp &&
+            DateUtils.isSameDay(
+              new Date(currentTimestamp),
+              new Date(timestamp),
+            ),
         );
+        if (index === sameDays.indexOf(true)) {
+          daySeparator = <DaySeparator date={new Date(currentTimestamp)} />;
+        }
       }
```

Comment:
It seems like you map every event to get the type, then loop to remove nulls, then loop to get all timestamps. Then loop over the events, and for each event loop the timestamps.

Would it be possible here to use reduce, loop once and return an object.

```
events.reduce((acc, curr) => {
const event = getValidTimelineEvent(curr);
if (!event) return acc;
// attach curr to acc and return
}, {})
```

having the structure like {'2018-06-01': { events: [...] }, '2018-06-02': { events : [...]}

Then loop over the keys in the object and between each key add the day separator.

Do you think this makes sense?

Reply:
Yes, that's a much cleaner way of doing it. I will give it a try. Thank you!